testbed/sampling/sampling_data_provider.go

```diff
@@ -0,0 +1,176 @@
+	return func(tc *samplingDataProvider) {
+		tc.shouldSample = shouldSample
+	}
+}
+
+// WithSampledSpanCustomizer allows you to customize the spans that are sampled.
+func WithSampledSpanCustomizer(customizer sampledSpanCustomizerFunc) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.sampledSpanCustomizer = customizer
+	}
+}
+
+// WithExpectedSampledTraces allows you to specify the number of spans per trace
+func WithSpansPerTrace(spansPerTrace int) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.spansPerTrace = spansPerTrace
+	}
+}
+
+// WithExpectedSampledTraces allows you to specify the total number of traces generated in a run
+func WithNumTracesToGenerate(numTracesToGenerate int) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.numTracesToGenerate = numTracesToGenerate
+	}
+}
+
+// NewSamplingDataProvider creates a new instance of samplingDataProvider
+func NewSamplingDataProvider(opts ...SamplingDataProviderOption) *samplingDataProvider {
+	sdp := samplingDataProvider{
+		// Defaults. These are arbitrary values.
+		spansPerTrace:       6,
+		numTracesToGenerate: 50,
+	}
+
+	for _, opt := range opts {
+		opt(&sdp)
+	}
+
+	return &sdp
+}
+
+// SetLoadGeneratorCounters sets the data providers items generated field to have a ref to the load generators counter
+func (sdp *samplingDataProvider) SetLoadGeneratorCounters(dataItemsGenerated *atomic.Uint64) {
+	sdp.dataItemsGenerated = dataItemsGenerated
+}
+
+// GenerateTraces the boolean in this return reflects whether we are done generating traces or not.
+func (sdp *samplingDataProvider) GenerateTraces() (ptrace.Traces, bool) {
+	if sdp.tracesGenerated >= sdp.numTracesToGenerate {
+		return ptrace.NewTraces(), true
+	}
+	sdp.tracesGenerated += 1
+
+	traceData := sdp.generateSampledTrace()
+
+	return traceData, false
+
+}
+
+func (sdp *samplingDataProvider) ToBeSampledSpansGenerated() uint64 {
+	return sdp.sampledSpansGenerated.Load()
+}
+
+func (sdp *samplingDataProvider) generateSampledTrace() ptrace.Traces {
+	traceData := ptrace.NewTraces()
+
+	spans := traceData.ResourceSpans().AppendEmpty().ScopeSpans().AppendEmpty().Spans()
+	spans.EnsureCapacity(sdp.spansPerTrace)
+
+	traceID := sdp.traceIDSequence.Add(1)
+	shouldSample := sdp.shouldSample(traceID)
+	// TODO: should we be using ItemsPerBatch to determine how many spans are in a trace?
+	for i := 0; i < sdp.spansPerTrace; i++ {
+
+		startTime := time.Now()
+		endTime := startTime.Add(time.Millisecond)
+
+		spanID := sdp.dataItemsGenerated.Add(1)
+
+		span := spans.AppendEmpty()
+
+		// Create a span.
+		span.SetTraceID(uInt64ToTraceID(0, traceID))
+		span.SetSpanID(uInt64ToSpanID(spanID))
+		span.SetName("dirty-trace-span")
+		span.SetKind(ptrace.SpanKindClient)
+		attrs := span.Attributes()
+
+		attrs.PutInt("dirty_trace.span_seq_num", int64(spanID))
+		attrs.PutInt("dirty_trace.trace_seq_num", int64(traceID))
```

Comment:
We should remove references to `dirty`. This was an initial implementation naming choice when I was going with `PollutedDataProvider`. It seems a little misnamed without that context :)